src/budget.rs

```diff
@@ -110,6 +117,175 @@ impl PaymentPlan for Budget {
+            } else {
+                accounts[0].tokens -= contract.tokens;
+            }
+        };
+    }
+    Ok(())
+}
+
+/// Apply only a transaction's credits.
+/// Note: It is safe to apply credits from multiple transactions in parallel.
+fn apply_credits_to_budget_state(
+    tx: &Transaction,
+    accounts: &mut [Account],
+    instruction: &Instruction,
+) -> Result<(), BudgetError> {
+    match instruction {
+        Instruction::NewContract(contract) => {
+            let plan = contract.plan.clone();
+            if let Some(payment) = plan.final_payment() {
+                accounts[1].tokens += payment.tokens;
+                Ok(())
+            } else if !accounts[1].userdata.is_empty() {
+                Err(BudgetError::ContractAlreadyPending(tx.keys[1]))
+            } else {
+                let mut state = BudgetState::default();
+                state.pending.insert(tx.signature, plan);
+                //TODO this is a temporary on demand allocation
+                //until system contract requires explicit allocation of memory
+                accounts[1].userdata = serialize(&state).unwrap();
+                accounts[1].tokens += contract.tokens;
+                Ok(())
+            }
+        }
+        Instruction::ApplyTimestamp(dt) => {
+            let mut state: BudgetState = deserialize(&accounts[1].userdata).unwrap();
+            state.apply_timestamp(tx.keys[0], *dt, &mut accounts[1]);
+            accounts[1].userdata = if state.is_pending() {
+                serialize(&state).unwrap()
+            } else {
+                vec![]
+            };
+            Ok(())
+        }
+        Instruction::ApplySignature(signature) => {
+            let mut state: BudgetState = deserialize(&accounts[1].userdata).unwrap();
+            state.apply_signature(tx.keys[0], *signature, accounts);
+            accounts[1].userdata = if state.is_pending() {
+                serialize(&state).unwrap()
+            } else {
+                vec![]
+            };
+            Ok(())
+        }
+        Instruction::NewVote(_vote) => {
+            // TODO: move vote instruction into a different contract
+            trace!("GOT VOTE! last_id={}", tx.last_id);
+            Ok(())
+        }
+    }
+}
+
+fn save_error_to_budget_state(e: BudgetError, accounts: &mut [Account]) -> () {
+    if !accounts[1].userdata.is_empty() {
+        let mut state: BudgetState = deserialize(&accounts[1].userdata).unwrap();
+        state.last_error = Some(e);
+        accounts[1].userdata = serialize(&state).unwrap();
+    }
+}
+
+/// Budget DSL contract interface
+/// * tx - the transaction
+/// * accounts[0] - The source of the tokens
+/// * accounts[1] - The contract context.  Once the contract has been completed, the tokens can
+/// be spent from this account .
+pub fn process_transaction(tx: &Transaction, accounts: &mut [Account]) -> () {
+    let instruction = deserialize(&tx.userdata).unwrap();
```

Comment:
worthwhile assert there are two accounts here?  I guess it'll panic in `apply_debits_to_budget_state` or `apply_credits_to_budget_state` if not...